mongosql/src/algebrizer/definitions.rs

```diff
@@ -904,32 +953,89 @@ impl<'a> Algebrizer<'a> {
     ) -> Result<mir::Stage> {
         *self.clause_type.borrow_mut() = ClauseType::Select;
-        match ast_node.set_quantifier {
-            ast::SetQuantifier::All => (),
-            ast::SetQuantifier::Distinct => return Err(Error::DistinctSelect),
-        };
 
-        match ast_node.body {
-            // Standard Select bodies must be only *, otherwise this is an
-            // error.
-            ast::SelectBody::Standard(exprs) => match exprs.as_slice() {
-                [ast::SelectExpression::Star] => {
-                    source.schema(&self.schema_inference_state())?;
-                    Ok(source)
+        match ast_node.set_quantifier {
+            ast::SetQuantifier::All => match ast_node.body {
+                // Standard Select bodies must be only *, otherwise this is an
+                // error.
+                ast::SelectBody::Standard(exprs) => match exprs.as_slice() {
+                    [ast::SelectExpression::Star] => {
+                        source.schema(&self.schema_inference_state())?;
+                        Ok(source)
+                    }
+                    _ => Err(Error::NonStarStandardSelectBody),
+                },
+                // SELECT VALUES expressions must be Substar expressions or normal Expressions that are
+                // Documents, i.e., that have Schema that Must satisfy ANY_DOCUMENT.
+                //
+                // All normal Expressions will be mapped as Datasource Bottom, and all Substars will be mapped
+                // as their name as a Datasource.
+                ast::SelectBody::Values(exprs) => {
+                    self.algebrize_select_values_body(exprs, source, is_add_fields, false)
+                }
+            },
+            ast::SetQuantifier::Distinct => match ast_node.body {
+                ast::SelectBody::Standard(exprs) => match exprs.as_slice() {
+                    [ast::SelectExpression::Star] => self.build_distinct_star_project_stage(source),
+                    _ => Err(Error::NonStarStandardSelectBody),
+                },
+                ast::SelectBody::Values(exprs) => {
+                    self.algebrize_select_values_body(exprs, source, is_add_fields, true)
                 }
-                _ => Err(Error::NonStarStandardSelectBody),
             },
-            // SELECT VALUES expressions must be Substar expressions or normal Expressions that are
-            // Documents, i.e., that have Schema that Must satisfy ANY_DOCUMENT.
-            //
-            // All normal Expressions will be mapped as Datasource Bottom, and all Substars will be mapped
-            // as their name as a Datasource.
-            ast::SelectBody::Values(exprs) => {
-                self.algebrize_select_values_body(exprs, source, is_add_fields)
-            }
         }
     }
 
+    fn build_distinct_star_project_stage(&self, source: mir::Stage) -> Result<mir::Stage> {
+        source.schema(&self.schema_inference_state())?;
```

Comment:
Is this a typo? We immediately do this again on line 992.

Reply:
It is, I removed it 👍